bitwarden_access_manager.py

```diff
@@ -0,0 +1,678 @@
+                    + ",\n]\n"
+                )
+            else:
+                result = result + "member_access = []\n"
+
+        if self.group_access is not None:
+            group_access_lines = [
+                "  " + a.format_toml() for a in sorted(self.group_access)
+            ]
+            if len(group_access_lines) > 0:
+                result = (
+                    result
+                    + "group_access = [\n"
+                    + ",\n".join(group_access_lines)
+                    + ",\n]"
+                )
+            else:
+                result = result + "group_access = []"
+        return result
+
+
+class BitwardenClient(NamedTuple):
+    connection: HTTPSConnection
+    bearer_token: str
+
+    @staticmethod
+    def new(client_id: str, client_secret: str) -> BitwardenClient:
+        response = requests.post(
+            "https://identity.bitwarden.com/connect/token",
+            headers={"Content-Type": "application/x-www-form-urlencoded"},
+            data={
+                "grant_type": "client_credentials",
+                "scope": "api.organization",
+                "Accept": "application/json",
+            },
+            auth=(client_id, client_secret),
+        )
+        bearer_token = response.json()["access_token"]
+        return BitwardenClient(HTTPSConnection("api.bitwarden.com"), bearer_token)
+
+    def _http_get(self, url: str) -> HTTPResponse:
+        self.connection.request(
+            method="GET",
+            url=url,
+            headers={
+                "Accept": "application/json",
+                "Authorization": f"Bearer {self.bearer_token}",
+            },
+        )
+
+        return self.connection.getresponse()
+
+    def get_groups(self) -> Iterable[Group]:
+        groups = json.load(self._http_get(f"/public/groups"))
+        for group in groups["data"]:
+            yield Group(
+                id=group["id"],
+                name=group["name"],
+                access_all=group["accessAll"],
+            )
+
+    def get_collection_members(
+        self, groups: Any, org_members: Dict[str, Member]
+    ) -> Iterable[MemberCollectionAccess]:
+        for group in groups:
+            group_id = group["id"]
+
+            member_ids = json.load(
+                self._http_get(f"/public/groups/{group_id}/member-ids")
+            )
+
+            for member_id in member_ids:
+                yield MemberCollectionAccess(
+                    name=org_members[member_id].name,
+                )
+
+    def get_collection_groups(self, groups: Any) -> Iterable[GroupCollectionAccess]:
+        for group in groups:
+            if group["readOnly"] == True:
+                access = "readonly"
+            else:
+                access = "write"
+
+            group_id = group["id"]
+            yield GroupCollectionAccess(
+                name=json.load(self._http_get(f"/public/groups/{group_id}"))["name"],
+                access=access,
+            )
```

Comment:
`GroupCollectionAccess` sounds to me like it controls which group can access which collections. But in this entire method, I see nothing related to collections. Doesn’t this just list all the groups that exist?

Reply:
`get_collection_groups` takes as argument a list of group. This is called by `get_collections` which retrieve all collections and pass as argument to `get_collection_groups` the API response containing an array of hash with the groups ids and their access to this collection [see](https://github.com/ChorusOne/github-access-manager/pull/10/files#diff-7bfcbcb708fa5299e4545d8f8335bc2ab16d37b5b7e70edd0262e41d040b133aR360). 

Output API endpoint `/public/collections/<collection_id>`:
```
{
  "externalId": "external_id_123456",
  "object": "collection",
  "id": "539a36c5-e0d2-4cf9-979e-51ecf5cf6593",
  "groups": [
    {
      "id": "bfbc8338-e329-4dc0-b0c9-317c2ebf1a09",
      "readOnly": true
    }
  ]
}
```